File: PLAYER/OMidiIn.cpp
//---------------------------------------------------------------------------
#ifdef USE_PRECOMPILED_HEADERS
#ifdef __BORLANDC__
#include "PrecompiledHeaders.h"
#pragma hdrstop
#endif
#endif

//---------------------------------------------------------------------------
#include "OMidiIn.h"
// -----------------------------------------------------------------------------
int NumeroDispositivosMIDIIn()
{
     return ( midiInGetNumDevs() );
};
// -----------------------------------------------------------------------------
char *ObtenerNombreDispositivoMidiIn(int Dispositivo)
{
    static MIDIINCAPS  M;
    static MMRESULT    FError;

    midiInGetDevCaps(Dispositivo,&M,sizeof(MIDIINCAPS));
    if (FError == MMSYSERR_NOERROR)
        return (M.szPname);
    else
        return(NULL);
};
// -----------------------------------------------------------------------------
TObjetoMidiIn::TObjetoMidiIn()
{
    Abierto = false;
}
// -----------------------------------------------------------------------------
void  TObjetoMidiIn::Open(void *FuncionCallBack)
{
    MMRESULT FError;
    DWORD    Flags;

    if (! Abierto)
    {
        if (FuncionCallBack==NULL)
          Flags=CALLBACK_NULL;
        else
          Flags=CALLBACK_FUNCTION;

        FError = midiInOpen((LPHMIDIIN)&Handler,          // se supone que nos lo devuelve
                           DispositivoActual,             // dispositivo a usar
                           (DWORD)FuncionCallBack,    // direccion de la RSI
                           (DWORD)0,                  // parametro de usuario
                           Flags);                        // Flags

        if (FError==MMSYSERR_NOERROR)
          Abierto=true;
    };
};
// -----------------------------------------------------------------------------
void  TObjetoMidiIn::Close()
{
     if (Abierto)
     {
          midiInClose(Handler);
          Abierto = false;
     }
}
// -----------------------------------------------------------------------------
void  TObjetoMidiIn::Start()
{
     if (Abierto) midiInStart(Handler);
}
// -----------------------------------------------------------------------------
void  TObjetoMidiIn::Stop()
{
     if (Abierto) midiInStop(Handler);
}
// -----------------------------------------------------------------------------
void  TObjetoMidiIn::EstablecerDispositivo(int D)
{
     if (! Abierto) DispositivoActual = D;
}
// -----------------------------------------------------------------------------
int TObjetoMidiIn::Dispositivo()
{
     return (DispositivoActual);
}
// -----------------------------------------------------------------------------

File: PLAYER/Omidiin.pas
unit OMidiIn;

interface
uses
    classes;
type
    TEventoMidiIn = RECORD
                          Cmd:byte;
                          Data1:byte;
                          Data2:byte;
                          Sysex:pointer;
                          TamSysex:longint;
                    END;

    TObjetoMidiIn = object
                          Abierto:Boolean;

                          constructor init;
                          procedure Open(FuncionCallBack:pointer);
                          procedure Close;
                          procedure Start;
                          procedure Stop;
                          procedure EstablecerDispositivo(D:integer);
                          function Dispositivo:Integer;

                          private
                          Handler:Integer;
                          DispositivoActual:Integer;
                    end;


    function NumeroDispositivosMIDIIn:Integer;
    procedure ObtenerListaDispositivosMIDIIn(S:TStrings);
    FUNCTION ObtenerNombreDispositivoMidiIn(Dispositivo:integer):String;

implementation
uses
    SysUtils,mmSystem;

function NumeroDispositivosMIDIIn:Integer;
begin
     NumeroDispositivosMIDIIn:=MidiInGetNumDevs;
end;

FUNCTION ObtenerNombreDispositivoMidiIn(Dispositivo:integer):String;
VAR
   M:TMidiInCaps;
   FError:Integer;
BEGIN
     FError :=midiInGetDevCaps(Dispositivo,@M,sizeof(TMidiInCaps));
     if (Ferror=MMSYSERR_NOERROR) then
        ObtenerNombreDispositivoMidiIn:=StrPas(M.szPName)
     else
         ObtenerNombreDispositivoMidiIn:='';
END;

procedure ObtenerListaDispositivosMIDIIn(S:TStrings);
var
   Ferror,j,i:integer;
   MidiInCaps:TMidiInCaps;
begin
     j:=MidiInGetNumDevs;

     S.Clear;
     for i:=0 to j-1 do
     begin
          FError :=midiInGetDevCaps(i,@MidiInCaps,sizeof(TMidiInCaps));
          S.Add(StrPas(MidiInCaps.szPname));
     end;
end;
{------------------------------------------------------------------------------}
constructor TObjetoMidiIn.init;
begin
     Abierto:=False;
end;

procedure TObjetoMidiIn.Open(FuncionCallBack:pointer);
VAR
   FError:INTEGER;
   Flags:longint;
begin
     IF (NOT Abierto) THEN
     BEGIN
          if (FuncionCallBack=nil) THEN Flags:=Callback_null else Flags:=Callback_Function;

          FError:=midiInOpen(@Handler,                      {se supone que nos lo devuelve}
                             DispositivoActual,             {dispositivo a usar }
                             Integer(FuncionCallBack),      {direccion de la RSI }
                             longint(0),                    {parametro de usuario}
                             Flags);                        {Flags               }

          IF (FError=MMSYSERR_NOERROR) then Abierto:=True;
     END;
end;

procedure TObjetoMidiIn.Close;
begin
     if Abierto then
     begin
          midiInClose(Handler);
          Abierto:=False;
     end;
end;

procedure TObjetoMidiIn.Start;
begin
     if Abierto then midiInStart(Handler);
end;

procedure TObjetoMidiIn.Stop;
begin
     if Abierto then midiInStop(Handler);
end;

procedure TObjetoMidiIn.EstablecerDispositivo(D:integer);
begin
     if (not Abierto) THEN DispositivoActual:=D;
end;

function TObjetoMidiIn.Dispositivo:Integer;
begin
     Dispositivo:=DispositivoActual;
end;
{------------------------------------------------------------------------------}
end.
